include/hip/hcc_detail/hip_vector_types.h

```diff
@@ -34,1132 +34,90 @@ THE SOFTWARE.
-    MAKE_COMPONENT_CONSTRUCTOR_TWO_COMPONENT(longlong2, unsigned int)
-    MAKE_COMPONENT_CONSTRUCTOR_TWO_COMPONENT(longlong2, signed int)
-    MAKE_COMPONENT_CONSTRUCTOR_TWO_COMPONENT(longlong2, float)
-    MAKE_COMPONENT_CONSTRUCTOR_TWO_COMPONENT(longlong2, double)
-    MAKE_COMPONENT_CONSTRUCTOR_TWO_COMPONENT(longlong2, unsigned long)
-    MAKE_COMPONENT_CONSTRUCTOR_TWO_COMPONENT(longlong2, signed long)
-    MAKE_COMPONENT_CONSTRUCTOR_TWO_COMPONENT(longlong2, unsigned long long)
-    MAKE_COMPONENT_CONSTRUCTOR_TWO_COMPONENT(longlong2, signed long long)
-#endif
-    signed long long x, y;
-} __attribute__((aligned(16)));
-
-struct longlong3 {
-#ifdef __cplusplus
-   public:
-    MAKE_DEFAULT_CONSTRUCTOR_THREE_COMPONENT(longlong3)
-
-    MAKE_COMPONENT_CONSTRUCTOR_THREE_COMPONENT(longlong3, unsigned char)
-    MAKE_COMPONENT_CONSTRUCTOR_THREE_COMPONENT(longlong3, signed char)
-    MAKE_COMPONENT_CONSTRUCTOR_THREE_COMPONENT(longlong3, unsigned short)
-    MAKE_COMPONENT_CONSTRUCTOR_THREE_COMPONENT(longlong3, signed short)
-    MAKE_COMPONENT_CONSTRUCTOR_THREE_COMPONENT(longlong3, unsigned int)
-    MAKE_COMPONENT_CONSTRUCTOR_THREE_COMPONENT(longlong3, signed int)
-    MAKE_COMPONENT_CONSTRUCTOR_THREE_COMPONENT(longlong3, float)
-    MAKE_COMPONENT_CONSTRUCTOR_THREE_COMPONENT(longlong3, double)
-    MAKE_COMPONENT_CONSTRUCTOR_THREE_COMPONENT(longlong3, unsigned long)
-    MAKE_COMPONENT_CONSTRUCTOR_THREE_COMPONENT(longlong3, signed long)
-    MAKE_COMPONENT_CONSTRUCTOR_THREE_COMPONENT(longlong3, unsigned long long)
-    MAKE_COMPONENT_CONSTRUCTOR_THREE_COMPONENT(longlong3, signed long long)
-#endif
-    signed long long x, y, z;
-};
-
-struct longlong4 {
-#ifdef __cplusplus
-   public:
-    MAKE_DEFAULT_CONSTRUCTOR_FOUR_COMPONENT(longlong4)
-
-    MAKE_COMPONENT_CONSTRUCTOR_FOUR_COMPONENT(longlong4, unsigned char)
-    MAKE_COMPONENT_CONSTRUCTOR_FOUR_COMPONENT(longlong4, signed char)
-    MAKE_COMPONENT_CONSTRUCTOR_FOUR_COMPONENT(longlong4, unsigned short)
-    MAKE_COMPONENT_CONSTRUCTOR_FOUR_COMPONENT(longlong4, signed short)
-    MAKE_COMPONENT_CONSTRUCTOR_FOUR_COMPONENT(longlong4, unsigned int)
-    MAKE_COMPONENT_CONSTRUCTOR_FOUR_COMPONENT(longlong4, signed int)
-    MAKE_COMPONENT_CONSTRUCTOR_FOUR_COMPONENT(longlong4, float)
-    MAKE_COMPONENT_CONSTRUCTOR_FOUR_COMPONENT(longlong4, double)
-    MAKE_COMPONENT_CONSTRUCTOR_FOUR_COMPONENT(longlong4, unsigned long)
-    MAKE_COMPONENT_CONSTRUCTOR_FOUR_COMPONENT(longlong4, signed long)
-    MAKE_COMPONENT_CONSTRUCTOR_FOUR_COMPONENT(longlong4, unsigned long long)
-    MAKE_COMPONENT_CONSTRUCTOR_FOUR_COMPONENT(longlong4, signed long long)
-#endif
-    signed long x, y, z, w;
-} __attribute__((aligned(32)));
+#if defined(__clang__)
+    #define __NATIVE_VECTOR__(n, ...) __attribute__((ext_vector_type(n)))
+#elif defined(__GNUC__) // N.B.: GCC does not support .xyzw syntax.
+    #define __NATIVE_VECTOR__(n, T) __attribute__((vector_size(n * sizeof(T))))
+#endif
+
```

Comment:
Essentially codes compiling with clang and gcc will be using different data types?   Will that affect extern functions that have vector types in its signature such that code compiling by one compiler can't call the function compiled by the other compiler due to name mangling difference?

Reply:
I believe that the only safe way to call across compilers is in extern "C", otherwise you are on your own anyway. Thus there'd be no mangling difference given that extern "C" functions are not mangled. Even if it were a mere struct there are no guarantees of mangling being matched between compilers (on the contrary). So whereas before that'd have been a quiet way to get yourself into trouble (oh, ok, I'm calling my super interesting template across ABI boundaries and it works yay... oh, snap, what just broke after this GCC upgrade?), this may fail early (annoyingly enough though, mangling is mostly matched at the moment).